# Step 3 error-weighted mean

Start with the first script you have, everything stays the same, except we add the
code to do the error-weighted mean. So you can skip down to cell 5

In [1]:
import numpy as np

In [2]:
data_file = np.loadtxt("test_data_lab01.csv",delimiter=',',skiprows=2)
v = data_file[:,0]
dv = data_file[:,1]
i = 0.001*data_file[:,2] 
di = 0.001*data_file[:,3]


In [3]:
def get_resistance_and_error(v,dv,i,di):
    #just Ohm's law
    resistance = v/i 
    #Use chain rule from calculus to calculate error
    error = np.sqrt(pow(dv/i,2) + pow(v*di/(i*i),2))
    return resistance, error

In [4]:
r =np.empty(len(v))
#initialise empty array to hold resistance uncertainty values
dr = np.empty(len(v))
for j in range(0, len(v)):
    r[j],dr[j]= get_resistance_and_error(v[j],dv[j],i[j],di[j])


In the cell below we do the error-weighted mean and print out the results

In [5]:
error_weighted_sum =0
error_weights=0    

#Do the sums
for j in range(0, len(v)):
    error_weighted_sum += r[j]/pow(dr[j],2)    
    error_weights += 1/pow(dr[j],2)
    
#Next calculate finalize the average weighting     
mean_resistance = error_weighted_sum/error_weights
mean_resistance_error = 1/np.sqrt(error_weights)

print("Weighted mean resistance = %.1f Ohm"%mean_resistance)
print("Standard error  = %.1f Ohm" %mean_resistance_error)

Weighted mean resistance = 221.3 Ohm
Standard error  = 4.5 Ohm


In a report you should quote the uncertainty to 1 significant figure so: 4.5 Ohms ---> 5 Ohms
Then you should quote the best estimate with matching precision. So it becomes 221 +/- 5 Ohms